src/main/java/sirius/biz/analytics/events/EventSpliterator.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Made with all the love in the world
+ * by scireum in Remshalden, Germany
+ *
+ * Copyright by scireum GmbH
+ * http://www.scireum.de - info@scireum.de
+ */
+
+package sirius.biz.analytics.events;
+
+import sirius.db.jdbc.OMA;
+import sirius.db.jdbc.SmartQuery;
+import sirius.kernel.async.TaskContext;
+import sirius.kernel.commons.PullBasedSpliterator;
+
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+/**
+ * Provides a spliterator which fetches events blockwise.
+ *
+ * @param <E> the type of events to fetch
+ */
+public class EventSpliterator<E extends Event<E>> extends PullBasedSpliterator<E> {
+
+    private static final int BLOCK_SIZE = 1000;
+
+    private final SmartQuery<E> query;
+
+    /**
+     * Contains the last events fetched (all sharing the same {@linkplain Event#getEventTimestamp() timestamp}) to
+     * prevent fetching the same events multiple times.
+     */
+    private List<E> lastEvents;
+
+    /**
+     * Contains a consumer which is used to prevent fetching the same events multiple times.
+     */
+    private final BiConsumer<SmartQuery<E>, E> duplicatePreventer;
+
+    /**
+     * Creates a new spliterator for the given query and duplicate preventer.
+     * <p>
+     * The duplicate preventer is supposed to add additional constraints to the query to prevent fetching the same
+     * events multiple times.
+     *
+     * @param query              the query to use to fetch the events
+     * @param duplicatePreventer a consumer which is used to prevent fetching the same events multiple times
+     */
+    public EventSpliterator(SmartQuery<E> query, BiConsumer<SmartQuery<E>, E> duplicatePreventer) {
+        super();
+        this.query = query.copy().orderAsc(Event.EVENT_TIMESTAMP).limit(BLOCK_SIZE);
+        this.duplicatePreventer = duplicatePreventer;
+    }
+
+    @Nullable
+    @Override
+    protected Iterator<E> pullNextBlock() {
+        if (!TaskContext.get().isActive()) {
+            return null;
+        }
+
+        if (lastEvents != null && lastEvents.size() < BLOCK_SIZE) {
+            return null;
+        }
+
+        List<E> events = resolveEffectiveQuery().queryList();
+        if (events.isEmpty()) {
+            return null;
+        }
+
+        lastEvents = events.reversed()
+                           .stream()
+                           .filter(event -> event.getEventTimestamp().equals(events.getLast().getEventTimestamp()))
```

Comment:
Couldn't this mean that the caller gets an empty intermediate iterator, even though more event with a later timestamp are available?
What if the number of events with the same timestamp exceeds the block size?

Reply:
Event Timestamps go up to a second. If we would change the blocks to be *around* 1000, we could just get away with timestamps. Basically, when we reach 1000 entries, go further until the last event on that second (yields more than 1000, but not millions). and the next query goes from that + 1 second.
A query in the OXOMI events shows less than 500 events within the same second.

Reply:
Correction: do have peaks > 1000, in this example with max 1664, which I think is quite OK in a list since events aren't that big. 
<img width="1559" alt="image" src="https://github.com/user-attachments/assets/155cb7f4-6be8-47db-98ba-05f0849b90af">

Reply:
My problem with this is that there are no guarantees at all

Reply:
Its OK. I just wanted to question the design as if we really need this duplicate preventer in the first place. My comment is perhaps not placed in the best line...